platform/darwin/src/MGLMapSnapshotter.mm

```diff
@@ -505,59 +449,70 @@ + (MGLImage*)drawAttributedSnapshotWorker:(mbgl::MapSnapshotter::Attributions)at
     dispatch_async(workQueue, ^{
-        // Call a class method to ensure we're not accidentally capturing self
-        MGLImage *compositedImage = [MGLMapSnapshotter drawAttributedSnapshotWorker:attributions
-                                                                      snapshotImage:mglImage
-                                                                         pointForFn:pointForFn
-                                                                        latLngForFn:latLngForFn
-                                                                              scale:scale
-                                                                               size:size
-                                                                          showsLogo:showsLogo
-                                                                     overlayHandler:overlayHandler];
+        // Call a function that cannot accidentally capture self.
+        MGLImage *compositedImage = MGLAttributedSnapshot(attributions, mglImage, options.scale, options.size, options.showsLogo, ^{
+            if (!overlayHandler) {
+                return;
+            }
+#if TARGET_OS_IPHONE
+            CGContextRef context = UIGraphicsGetCurrentContext();
+            if (!context) {
+                return;
+            }
+            MGLMapSnapshotOverlay *snapshotOverlay = [[MGLMapSnapshotOverlay alloc] initWithContext:context
+                                                                                              scale:options.scale
+                                                                                         pointForFn:pointForFn
+                                                                                        latLngForFn:latLngForFn];
+            CGContextSaveGState(context);
+            overlayHandler(snapshotOverlay);
+            CGContextRestoreGState(context);
+#else
+            NSGraphicsContext *context = [NSGraphicsContext currentContext];
+            if (!context) {
+                return;
+            }
+            MGLMapSnapshotOverlay *snapshotOverlay = [[MGLMapSnapshotOverlay alloc] initWithContext:context.CGContext
+                                                                                              scale:options.scale
+                                                                                         pointForFn:pointForFn
+                                                                                        latLngForFn:latLngForFn];
+            [context saveGraphicsState];
+            overlayHandler(snapshotOverlay);
+            [context restoreGraphicsState];
+#endif
+        });
 
         // Dispatch result to origin queue
         dispatch_async(resultQueue, ^{
-            __typeof__(self) strongself = weakself;
-
-            if (strongself.completion) {
-
+            if (completion) {
+                __typeof__(self) strongSelf = weakSelf;
+                strongSelf.loading = NO;
```

Comment:
This is happening on the result queue, but `loading` could’ve been set to `YES` on a different queue.